contracts/pow/pow.cpp

```diff
@@ -23,70 +29,185 @@ struct pow_signature_data
 
-struct difficulty_metadata
+struct difficulty_metadata_v1
 {
    uint256_t      difficulty_target = 0;
    timestamp_type last_block_time = timestamp_type( 0 );
    timestamp_type block_window_time = timestamp_type( 0 );
    uint32_t       averaging_window = 0;
 };
 
-KOINOS_REFLECT( difficulty_metadata,
+KOINOS_REFLECT( difficulty_metadata_v1,
    (difficulty_target)
    (last_block_time)
    (block_window_time)
    (averaging_window)
 )
 
-difficulty_metadata get_difficulty_meta()
+struct difficulty_metadata_v2
 {
-   difficulty_metadata diff_meta;
-   system::db_get_object( contract_id, DIFFICULTY_METADATA_KEY, diff_meta );
-   if ( diff_meta.difficulty_target == 0 )
+   uint256_t      difficulty_target = 0;
+   timestamp_type last_block_time = timestamp_type( 0 );
+   uint64_t       required_hc = 0;
+   uint128_t      hc_reserve;
+   uint128_t      us_reserve;
+   timestamp_type target_block_interval = timestamp_type( TARGET_BLOCK_INTERVAL_MS / 1000 );
+};
+
+KOINOS_REFLECT( difficulty_metadata_v2,
+   (difficulty_target)
+   (last_block_time)
+   (required_hc)
+   (hc_reserve)
+   (us_reserve)
+   (target_block_interval)
+)
+
+using difficulty_metadata = difficulty_metadata_v2;
+
+uint64_t get_hc_reserve_multiplier(uint32_t effective_us)
+{
+   // Rescale so that 2^32 represents one half-life
+   uint64_t udt = (uint64_t( effective_us ) << 32) / (uint64_t(1000) * HC_HALFLIFE_MS);
+   int64_t idt = int64_t(udt);
+
+   int64_t y = -0xa2b23f3;
+   y *= idt;
+   y >>= 32;
+   y += 0x3b9d3bec;
+   y *= idt;
+   y >>= 32;
+   y -= 0xb17217f7;
+   y *= idt;
+
+   y >>= 32;
+   y += 0x100000000;
+   if( y < 0 )
+      y = 0;
+   return uint64_t( y );
+   /*
+   y = (y > 0) ? 0 : -y;
+   return uint64_t( y );
+   */
+}
+
+void set_difficulty_target( difficulty_metadata& diff_meta )
+{
+   if( diff_meta.us_reserve <= TARGET_BLOCK_INTERVAL_MS )
    {
-      diff_meta.difficulty_target = std::numeric_limits< uint256_t >::max() >> 26;
+      diff_meta.difficulty_target = 0;
+      return;
    }
 
+   // x*y = k
+   // -> (x+dx)*(y+dy) = x*y
+   // -> x*y + dx*y + dy*x + dx*dy = x*y
+   // -> dx*(y+dy) = -dy*x
+   // -> dx = -dy*x/(y+dy)
+   //
+   // dy = -TARGET_BLOCK_INTERVAL_MS
+
+   // TODO:  Overflow analysis
+
+   diff_meta.required_hc = uint64_t( ((1000 * TARGET_BLOCK_INTERVAL_MS) * diff_meta.hc_reserve) / (diff_meta.us_reserve - (1000 * TARGET_BLOCK_INTERVAL_MS)) );
+   ++diff_meta.required_hc;
+   diff_meta.difficulty_target = std::numeric_limits< uint256_t >::max() / diff_meta.required_hc;
+}
+
+difficulty_metadata initialize_difficulty( const difficulty_metadata_v1& pow1_meta )
+{
+   uint64_t initial_hashes_per_block = ( std::numeric_limits< uint256_t >::max() / pow1_meta.difficulty_target ).convert_to< uint64_t >();
+   uint64_t initial_hashes_per_second = initial_hashes_per_block / ( TARGET_BLOCK_INTERVAL_MS / 1000 );
+   uint64_t effective_us = uint64_t( 1000 ) * TARGET_BLOCK_INTERVAL_MS;
```

Comment:
This computation should be `initial_hashes_per_second = (1000 * initial_hashes_per_block) / TARGET_BLOCK_INTERVAL_MS;` so it Does the Right Thing even if `TARGET_BLOCK_INTERVAL_MS` is not evenly divisible by 1000.

However, if we do this, we also need an overflow check to be sure the multiplication in the numerator doesn't overflow.  Which can go in the same overflow analysis TODO issue as other overflow checks in this smart contract.